File: backend/app/schemas/action_center.py
import uuid
from datetime import datetime, date
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, ConfigDict


class AvailableActionOption(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    action: str  # APPROVE, REJECT, REQUEST_REWORK, FORWARD, ASSIGN, VERIFY, SUBMIT, RECORD_POSSESSION, PROCESS, REVIEW_PAYMENT, UPDATE_CASE, COMPLETE, RESPOND_REWORK
    label: str
    description: str
    permission: str
    target_status: Optional[str] = None
    requires_remarks: bool = False
    requires_rejection_reason: bool = False
    requires_document: bool = False
    requires_confirmation: bool = True
    is_primary: bool = False
    badge_variant: str = "default"  # default, primary, danger, warning, outline
    target_authority_options: Optional[List[Dict[str, str]]] = None


class ActionItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str  # task UUID or synthetic ID
    action_type: str  # PROJECT_SCRUTINY, FIELD_VERIFICATION, OBJECTION_REVIEW, COMPENSATION_ASSESSMENT, AWARD_ACTION, POSSESSION_EXECUTION, RANDR_REVIEW, ESCALATION, DOCUMENT_REVIEW, CLARIFICATION
    title: str
    description: Optional[str] = None
    project_id: Optional[uuid.UUID] = None
    project_code: Optional[str] = None
    project_title: Optional[str] = None
    record_type: str  # Project, LandParcel, FieldVerification, AffectedFamily, CompensationAssessment, Award, Possession, Document, Escalation
    record_reference: str  # e.g., NH-48-EXP-2024, KH-104/2, AF-2024-JPR-001, DOC-DPR-01
    workflow_stage: str
    workflow_stage_name: str
    status: str  # PENDING, IN_PROGRESS, REWORK_REQUIRED, FORWARDED, COMPLETED
    priority: str  # CRITICAL, HIGH, NORMAL, LOW
    due_date: Optional[date] = None
    sla_days_remaining: Optional[int] = None
    is_overdue: bool = False
    assigned_role: str
    assigned_role_name: str
    assigned_user_id: Optional[uuid.UUID] = None
    assigned_user_name: Optional[str] = None
    required_action_summary: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    rework_reason: Optional[str] = None
    forwarded_to_role: Optional[str] = None
    forwarded_to_user_name: Optional[str] = None
    primary_action: Optional[AvailableActionOption] = None


class ActionCenterKpis(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    requires_action: int = 0
    due_soon: int = 0
    overdue: int = 0
    high_priority: int = 0
    in_progress: int = 0
    returned_rework: int = 0
    forwarded: int = 0
    completed: int = 0


class ActionCenterSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    kpis: ActionCenterKpis
    my_actions: List[ActionItemResponse] = []
    in_progress: List[ActionItemResponse] = []
    returned_rework: List[ActionItemResponse] = []
    forwarded: List[ActionItemResponse] = []
    completed: List[ActionItemResponse] = []
    recent_activity: List[Dict[str, Any]] = []


class RequiredActionDetail(BaseModel):
    what_needs_to_be_done: str
    why_it_is_required: str
    information_or_documents_needed: List[str] = []
    what_happens_after_completion: str
    statutory_reference: Optional[str] = None


class RecordInformationSection(BaseModel):
    record_type: str
    record_reference: str
    title: str
    key_attributes: Dict[str, Any] = {}
    financial_details: Optional[Dict[str, Any]] = None
    geographic_details: Optional[Dict[str, Any]] = None
    beneficiary_details: Optional[Dict[str, Any]] = None


class ActionDocumentItem(BaseModel):
    id: str
    document_name: str
    category: str
    version_number: int
    is_current: bool
    uploaded_by: str
    uploaded_at: datetime
    file_size_bytes: Optional[int] = None
    verification_status: str  # PENDING, ACCEPTED, REJECTED, REWORK
    file_url: Optional[str] = None
    sha256_hash: Optional[str] = None


class ActionRemarkItem(BaseModel):
    id: str
    author_name: str
    author_role: str
    action: str
    remarks: str
    timestamp: datetime


class ActionTimelineEvent(BaseModel):
    stage_name: str
    decision: str
    officer_name: str
    officer_role: str
    remarks: Optional[str] = None
    timestamp: datetime


class ActionWorkspaceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    action_item: ActionItemResponse
    case_summary: Dict[str, Any]
    required_action: RequiredActionDetail
    record_information: RecordInformationSection
    available_actions: List[AvailableActionOption]
    remarks_history: List[ActionRemarkItem] = []
    documents: List[ActionDocumentItem] = []
    workflow_timeline: List[ActionTimelineEvent] = []
    audit_history: List[Dict[str, Any]] = []


class ActionExecuteRequest(BaseModel):
    action: str  # APPROVE, REJECT, REQUEST_REWORK, FORWARD, ASSIGN, VERIFY, SUBMIT, RECORD_POSSESSION, PROCESS, REVIEW_PAYMENT, UPDATE_CASE, COMPLETE, RESPOND_REWORK
    remarks: Optional[str] = None
    rejection_reason: Optional[str] = None
    rework_reason: Optional[str] = None
    rework_items: Optional[List[str]] = None
    target_authority_role: Optional[str] = None
    target_user_id: Optional[uuid.UUID] = None
    due_date: Optional[date] = None
    document_ids: Optional[List[str]] = None
    extra_payload: Optional[Dict[str, Any]] = None


class ActionAssignRequest(BaseModel):
    action_id: str
    target_role: str
    target_user_id: Optional[uuid.UUID] = None
    instructions: str
    priority: str = "NORMAL"
    due_date: Optional[date] = None


class ActionForwardRequest(BaseModel):
    target_role: str
    target_user_id: Optional[uuid.UUID] = None
    forwarding_remarks: str
    supporting_document_ids: Optional[List[str]] = None


class ActionReworkRequest(BaseModel):
    rework_reason: str
    rework_items: List[str] = []
    due_date: Optional[date] = None
    remarks: Optional[str] = None

File: frontend/src/lib/types/action_center.ts
export interface AvailableActionOption {
  action: string;
  label: string;
  description: string;
  permission: string;
  target_status?: string;
  requires_remarks: boolean;
  requires_rejection_reason: boolean;
  requires_document: boolean;
  requires_confirmation: boolean;
  is_primary: boolean;
  badge_variant: "default" | "primary" | "danger" | "warning" | "outline";
  target_authority_options?: { role: string; label: string }[];
}

export interface ActionItemResponse {
  id: string;
  action_type: string;
  title: string;
  description?: string;
  project_id?: string;
  project_code?: string;
  project_title?: string;
  record_type: string;
  record_reference: string;
  workflow_stage: string;
  workflow_stage_name: string;
  status: "PENDING" | "IN_PROGRESS" | "REWORK_REQUIRED" | "FORWARDED" | "COMPLETED" | "SUBMITTED" | "REJECTED";
  priority: "CRITICAL" | "HIGH" | "NORMAL" | "LOW";
  due_date?: string;
  sla_days_remaining?: number;
  is_overdue: boolean;
  assigned_role: string;
  assigned_role_name: string;
  assigned_user_id?: string;
  assigned_user_name?: string;
  required_action_summary: string;
  created_at: string;
  updated_at?: string;
  rework_reason?: string;
  forwarded_to_role?: string;
  forwarded_to_user_name?: string;
  primary_action?: AvailableActionOption;
}

export interface ActionCenterKpis {
  requires_action: number;
  due_soon: number;
  overdue: number;
  high_priority: number;
  in_progress: number;
  returned_rework: number;
  forwarded: number;
  completed: number;
}

export interface ActionCenterSummaryResponse {
  kpis: ActionCenterKpis;
  my_actions: ActionItemResponse[];
  in_progress: ActionItemResponse[];
  returned_rework: ActionItemResponse[];
  forwarded: ActionItemResponse[];
  completed: ActionItemResponse[];
  recent_activity: {
    id: string;
    action: string;
    title: string;
    project: string;
    user: string;
    role: string;
    timestamp: string;
    details: string;
  }[];
}

export interface RequiredActionDetail {
  what_needs_to_be_done: string;
  why_it_is_required: string;
  information_or_documents_needed: string[];
  what_happens_after_completion: string;
  statutory_reference?: string;
}

export interface RecordInformationSection {
  record_type: string;
  record_reference: string;
  title: string;
  key_attributes: Record<string, any>;
  financial_details?: Record<string, any>;
  geographic_details?: Record<string, any>;
  beneficiary_details?: Record<string, any>;
}

export interface ActionDocumentItem {
  id: string;
  document_name: string;
  category: string;
  version_number: number;
  is_current: boolean;
  uploaded_by: string;
  uploaded_at: string;
  file_size_bytes?: number;
  verification_status: "PENDING" | "ACCEPTED" | "REJECTED" | "REWORK";
  file_url?: string;
  sha256_hash?: string;
}

export interface ActionRemarkItem {
  id: string;
  author_name: string;
  author_role: string;
  action: string;
  remarks: string;
  timestamp: string;
}

export interface ActionTimelineEvent {
  stage_name: string;
  decision: string;
  officer_name: string;
  officer_role: string;
  remarks?: string;
  timestamp: string;
}

export interface ActionWorkspaceResponse {
  action_item: ActionItemResponse;
  case_summary: Record<string, any>;
  required_action: RequiredActionDetail;
  record_information: RecordInformationSection;
  available_actions: AvailableActionOption[];
  remarks_history: ActionRemarkItem[];
  documents: ActionDocumentItem[];
  workflow_timeline: ActionTimelineEvent[];
  audit_history: {
    action: string;
    user: string;
    role: string;
    timestamp: string;
    details: string;
  }[];
}

export interface ActionExecuteRequest {
  action: string;
  remarks?: string;
  rejection_reason?: string;
  rework_reason?: string;
  rework_items?: string[];
  target_authority_role?: string;
  target_user_id?: string;
  due_date?: string;
  document_ids?: string[];
  extra_payload?: Record<string, any>;
}

export interface ActionAssignRequest {
  action_id: string;
  target_role: string;
  target_user_id?: string;
  instructions: string;
  priority?: string;
  due_date?: string;
}

export interface ActionForwardRequest {
  target_role: string;
  target_user_id?: string;
  forwarding_remarks: string;
  supporting_document_ids?: string[];
}

export interface ActionReworkRequest {
  rework_reason: string;
  rework_items?: string[];
  due_date?: string;
  remarks?: string;
}
